keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,385 @@
+
+```go
+
+// FailurePolicyAction defines the action the JobSet controller will take for
+// a given FailurePolicyRule.
+type FailurePolicyAction string
+
+const (
+	// Fail JobSet immediately, regardless of maxRestarts.
+	FailJobSet FailurePolicyAction = "FailJobSet"
+
+	// Don't count the failure against maxRestarts.
+	Ignore FailurePolicyAction = "Ignore"
+
+	// The failed child job is permanently failed and should not be
+	// restarted, active child jobs will continue to run to completion.
+	FailJob FailurePolicyAction = "FailJob"
+
+	// Restart the failed child job only, not the whole JobSet.
+	RestartJob FailurePolicyAction = "RestartJob"
+)
+
+// FailurePolicyRule defines a FailurePolicyAction to be executed if a child job
+// fails due to a reason listed in OnJobFailureReasons.
+// Only only rule can match a unique (Job failure reason, ReplicatedJob) pair.
+type FailurePolicyRule struct {
+	// The action to take if the rule is matched.
+	Action FailurePolicyAction `json:"action"`
+	// The requirement on the job failure reasons. The requirement
+	// is satisfied if at least one reason matches the list.
+	// Each unique failure reason can only be associated with one FailurePolicyRule.
+  // An empty list applies the rule to any job failure reason.
+	OnJobFailureReasons []string `json:"onJobFailureReasons"`
+	// TargetReplicatedJobs are the names of the replicated jobs the operator will apply to.
+	// A null or empty list will apply to all replicatedJobs.
+	// +optional
+	// +listType=atomic
+	TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
+}
+
+type FailurePolicy struct {
+	// MaxRestarts defines the limit on the number of JobSet restarts.
+	// A restart is achieved by recreating all active child jobs.
+	MaxRestarts int32 `json:"maxRestarts,omitempty"`
+	// List of failure policy rules for this JobSet.
+  // Only one rule can match a unique (Job failure reason, ReplicatedJob)
+  // pair. If such a rule exists, it is executed. Otherwise, the default
+  // failure policy of attempting to restart the JobSet is applied.
+	Rules []FailurePolicyRule `json:"rules,omitempty"`
+}
+
+```
+
+### Constraints
+
+- Each unique (Job failure reason, target Replicated Job) pair can only be associated with one FailurePolicyRule.
+This gives the user the flexibility to specify different failure and restart semantics for jobs failing for the same
+reason, depending on what Replicated Job they are part of. However, within a Replicated Job, there cannot be more than
+one way of handling a particular Job failure reason.
+
+- OnJobFailureReasons must be a valid Job failure reason as defined [here](https://github.com/kubernetes/kubernetes/blob/2d4100335e4c4ccc28f96fac78153f378212da4c/staging/src/k8s.io/api/batch/v1/types.go#L537-L554)
+in the batchv1 Job API. At the time of writing this KEP, these include:
+  - `PodFailurePolicy`
+  - `BackoffLimitExceeded`
+  - `DeadlineExceeded`
+  - `MaxFailedIndexesExceeded`
```

Comment:
These are alpha features so maybe we leave those off for now.

We should also think about SuccessPolicy in Job if that feature gets accepted.

Reply:
> These are alpha features so maybe we leave those off for now.

Looks like they are in beta now, as of 1.29.

Reply:
But these will be arbitrary strings in the future.